cluster/tls_transport.go

```diff
@@ -0,0 +1,335 @@
+	}
+	addr := &net.TCPAddr{IP: ip, Port: bindPort}
+	listener, err := tls.Listen(network, addr.String(), tlsConf)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to start TLS listener on %q port %d", bindAddr, bindPort))
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	t := &TLSTransport{
+		ctx:      ctx,
+		cancel:   cancel,
+		logger:   logger,
+		bindAddr: bindAddr,
+		bindPort: bindPort,
+		done:     make(chan struct{}),
+		handlers: handlers,
+		listener: listener,
+		packetCh: make(chan *memberlist.Packet),
+		streamCh: make(chan net.Conn),
+		tlsConf:  tlsConf,
+	}
+
+	t.registerMetrics(reg)
+
+	go func() {
+		t.listen()
+		close(t.done)
+	}()
+	return t, nil
+}
+
+// FinalAdvertiseAddr is given the user's configured values (which
+// might be empty) and returns the desired IP and port to advertise to
+// the rest of the cluster.
+func (t *TLSTransport) FinalAdvertiseAddr(ip string, port int) (net.IP, int, error) {
+	var advertiseAddr net.IP
+	var advertisePort int
+	if ip != "" {
+		advertiseAddr = net.ParseIP(ip)
+		if advertiseAddr == nil {
+			return nil, 0, fmt.Errorf("failed to parse advertise address %q", ip)
+		}
+
+		if ip4 := advertiseAddr.To4(); ip4 != nil {
+			advertiseAddr = ip4
+		}
+		advertisePort = port
+	} else {
+		if t.bindAddr == "0.0.0.0" {
+			// Otherwise, if we're not bound to a specific IP, let's
+			// use a suitable private IP address.
+			var err error
+			ip, err = sockaddr.GetPrivateIP()
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to get interface addresses: %v", err)
+			}
+			if ip == "" {
+				return nil, 0, fmt.Errorf("no private IP address found, and explicit IP not provided")
+			}
+
+			advertiseAddr = net.ParseIP(ip)
+			if advertiseAddr == nil {
+				return nil, 0, fmt.Errorf("failed to parse advertise address: %q", ip)
+			}
+		} else {
+			advertiseAddr = t.listener.Addr().(*net.TCPAddr).IP
+		}
+		advertisePort = t.GetAutoBindPort()
+	}
+	return advertiseAddr, advertisePort, nil
+}
+
+// PacketCh returns a channel that can be read to receive incoming
+// packets from other peers.
+func (t *TLSTransport) PacketCh() <-chan *memberlist.Packet {
+	return t.packetCh
+}
+
+// StreamCh returns a channel that can be read to handle incoming stream
+// connections from other peers.
+func (t *TLSTransport) StreamCh() <-chan net.Conn {
+	return t.streamCh
+}
+
+// Shutdown is called when memberlist is shutting down; this gives the
+// TLS Transport a chance to clean up the listener and other goroutines.
+func (t *TLSTransport) Shutdown() error {
+	level.Debug(t.logger).Log("msg", "shutting down tls transport")
+	t.cancel()
+	err := t.listener.Close()
+	<-t.done
+	return err
+}
+
+// WriteTo is a packet-oriented interface that creates a connection,
+// writes to it, and closes it. It also returns a time stamp of when
+// the packet was written.
+func (t *TLSTransport) WriteTo(b []byte, addr string) (time.Time, error) {
+	dialer := &net.Dialer{Timeout: DefaultTcpTimeout}
```

Comment:
Do I understand correctly, that this is opening up a new TCP connection for each packet to send? What do you think of something along the lines of a [connection pool](https://github.com/mxinden/memberlist-tls-transport/blob/master/tls_transport.go#L178)?

Reply:
It adds additional complexity as we need to manage the lifecycle of the connection and the Memberlist does not provide any good way to handle it. I would not recommend using a connection pool in this case. 
In our testing, we did not see any issues yet. If you like we can run additional load tests to identify if we need a pool or not.

Reply:
What sort of network latency did the testing have?

Reply:
An individual request is taking less than 3 mill seconds. Here are the benchmark results
goos: darwin
goarch: amd64
pkg: github.com/prometheus/alertmanager/cluster
BenchmarkWriteTo-12    	1000000000	         0.00304 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/prometheus/alertmanager/cluster	0.111s
Success: Benchmarks passed.

Reply:
The question isn't about how long things take over localhost, it's how many round trips may be required over a 200ms+ connection.

Reply:
Oh got it. I will run a test against remotely deployed Alertmanger next week. If we see any issue we will implement a connection pool.
Thank you.

Reply:
We've added a connection pool :)